### PR TITLE
Shopnuts

### PR DESCRIPTION
**Pull request by gamestabled**

### CODE REVIEW

File: src/overlays/actors/ovl_En_Hintnuts/z_en_hintnuts.c [hunk 1]
```diff
@@ -220,38 +220,27 @@ void EnHintnuts_SetupFreeze(EnHintnuts* this) {
...
-    f32 boundedCurrentFrameTemp;
 
     hasSlowPlaybackSpeed = 0;
     if (this->skelAnime.playSpeed < 0.5f) {
         hasSlowPlaybackSpeed = 1;
     }
-    if (hasSlowPlaybackSpeed && this->animFlagAndTimer != 0) {
-        this->animFlagAndTimer--;
+    if (hasSlowPlaybackSpeed) {
+        DECR(this->animFlagAndTimer);
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Hintnuts/z_en_hintnuts.c:229]: I don't think we're doing `DECR` outside of if statements.

File: src/overlays/actors/ovl_En_Hintnuts/z_en_hintnuts.c [hunk 2]
```diff
@@ -220,38 +220,27 @@ void EnHintnuts_SetupFreeze(EnHintnuts* this) {
...
-        if (this->skelAnime.curFrame > 12.0f) {
-            boundedCurrentFrameTemp = 12.0f;
-        } else {
-            boundedCurrentFrameTemp = this->skelAnime.curFrame;
-        }
-        boundedCurrentFrame = boundedCurrentFrameTemp;
-    }
-    this->collider.dim.height = (((boundedCurrentFrame - 9.0f) * 9.0f) + 5.0f);
+
+    this->collider.dim.height = (((CLAMP(this->skelAnime.curFrame, 9.0f, 12.0f) - 9.0f) * 9.0f) + 5.0f);
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Hintnuts/z_en_hintnuts.c:237]:
```suggestion
    this->collider.dim.height = 5.0f + ((CLAMP(this->skelAnime.curFrame, 9.0f, 12.0f) - 9.0f) * 9.0f);
```
Might be more clear this way, assuming it matches. Either way drop the outermost `()`.

File: src/overlays/actors/ovl_En_Hintnuts/z_en_hintnuts.c [hunk 3]
```diff
@@ -308,14 +297,10 @@ void EnHintnuts_ThrowNut(EnHintnuts* this, GlobalContext* globalCtx) {
...
 void EnHintnuts_Burrow(EnHintnuts* this, GlobalContext* globalCtx) {
-    f32 boundedCurrentFrame;
-
     if (SkelAnime_Update(&this->skelAnime) != 0) {
         EnHintnuts_SetupWait(this);
     } else {
-        boundedCurrentFrame =
-            this->skelAnime.curFrame < 1.0f ? 1.0f : this->skelAnime.curFrame > 3.0f ? 3.0f : this->skelAnime.curFrame;
-        this->collider.dim.height = (((3.0f - boundedCurrentFrame) * 12.0f) + 5.0f);
+        this->collider.dim.height = (((3.0f - CLAMP(this->skelAnime.curFrame, 1.0f, 3.0f)) * 12.0f) + 5.0f);
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Hintnuts/z_en_hintnuts.c:303]: same as before

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 4]
```diff
@@ -30,41 +37,265 @@ const ActorInit En_Shopnuts_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Shopnuts/EnShopnuts_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Shopnuts/EnShopnuts_Destroy.s")
+static ColliderCylinderInit sCylinderInit = {
+    { COLTYPE_UNK6, 0x00, 0x09, 0x39, 0x10, COLSHAPE_CYLINDER },
+    { 0x00, { 0x00000000, 0x00, 0x00 }, { 0xFFCFFFFF, 0x00, 0x00 }, 0x00, 0x01, 0x01 },
+    { 0x0014, 0x0028, 0x0000, { 0x0000, 0x0000, 0x0000 } }
+};
+
+static CollisionCheckInfoInit sColChkInfoInit = { 0x01, 0x0014, 0x0028, 0xFE };
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:47]:
```suggestion
static CollisionCheckInfoInit sColChkInfoInit = { 1, 20, 40, 0xFE };
```
These should be in decimal, with an exception for `0xFE` and `0xFF` in the last slot.

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 5]
```diff
@@ -30,41 +37,265 @@ const ActorInit En_Shopnuts_InitVars = {
...
+
+static InitChainEntry sInitChain[] = { ICHAIN_S8(naviEnemyId, 0x4E, ICHAIN_CONTINUE),
+                                       ICHAIN_F32(gravity, -1, ICHAIN_CONTINUE),
+                                       ICHAIN_F32(unk_4C, 2600, ICHAIN_STOP) };
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:51]:
```suggestion
                                       ICHAIN_F32(unk_4C, 2600, ICHAIN_STOP), };
```

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 6]
```diff
@@ -30,41 +37,265 @@ const ActorInit En_Shopnuts_InitVars = {
...
+
+void EnShopnuts_Wait(EnShopnuts* this, GlobalContext* globalCtx) {
+    s32 hasSlowPlaybackSpeed;
+
+    hasSlowPlaybackSpeed = 0;
+    if (this->skelAnime.playSpeed < 0.5f) {
+        hasSlowPlaybackSpeed = 1;
+    }
+    if (hasSlowPlaybackSpeed) {
+        DECR(this->animFlagAndTimer);
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:130]:
```suggestion
        DECR(this->animFlagAndTimer);
```
another DECR

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 7]
```diff
@@ -30,41 +37,265 @@ const ActorInit En_Shopnuts_InitVars = {
...
+    }
+    if (Animation_OnFrame(&this->skelAnime, 9.0f) != 0) {
+        this->collider.base.acFlags |= 1;
+    } else if (Animation_OnFrame(&this->skelAnime, 8.0f) != 0) {
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_NUTS_UP);
+    }
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:136]:
```suggestion
    if (Animation_OnFrame(&this->skelAnime, 9.0f)) {
        this->collider.base.acFlags |= 1;
    } else if (Animation_OnFrame(&this->skelAnime, 8.0f)) {
        Audio_PlayActorSound2(&this->actor, NA_SE_EN_NUTS_UP);
    }
```
`Animation_OnFrame` returns a boolean.

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 8]
```diff
@@ -30,41 +37,265 @@ const ActorInit En_Shopnuts_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Shopnuts/func_80AFA880.s")
+    this->collider.dim.height = ((CLAMP(this->skelAnime.curFrame, 9.0f, 13.0f) - 9.0f) * 9.0f) + 5.0f;
+    if (!hasSlowPlaybackSpeed && (this->actor.xzDistFromLink < 120.0f)) {
+        EnShopnuts_SetupBurrow(this);
+    } else if (SkelAnime_Update(&this->skelAnime) != 0) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:141]:
```suggestion
    } else if (SkelAnime_Update(&this->skelAnime)) {
```
as does `SkelAnime_Update`

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 9]
```diff
@@ -30,41 +37,265 @@ const ActorInit En_Shopnuts_InitVars = {
...
+        Actor_Kill(&this->actor);
+    } else {
+        Math_ApproachS(&this->actor.shape.rot.y, this->actor.yawTowardsLink, 2, 0xE38);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Shopnuts/func_80AFACE0.s")
+void EnShopnuts_ColliderCheck(EnShopnuts* this, GlobalContext* globalCtx) {
+    if (this->collider.base.acFlags & 0x2) {
+        this->collider.base.acFlags &= ~0x2;
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:225]:
```suggestion
    if (this->collider.base.acFlags & 2) {
        this->collider.base.acFlags &= ~2;
```
To make things easier for when collision_check gets updated, don't put the `0x` on these single-digit flags.

**Review by petrie911 — COMMENTED** [review]
Looks good. Just a few formatting comments you should handle.

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 10]
```diff
@@ -30,41 +37,266 @@ const ActorInit En_Shopnuts_InitVars = {
...
+    this->collider.base.acFlags &= ~1;
+    this->actionFunc = EnShopnuts_SpawnSalesman;
+}
+
+void EnShopnuts_Wait(EnShopnuts* this, GlobalContext* globalCtx) {
+    s32 hasSlowPlaybackSpeed;
+
+    hasSlowPlaybackSpeed = 0;
+    if (this->skelAnime.playSpeed < 0.5f) {
+        hasSlowPlaybackSpeed = 1;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:128]: can use true and false for the bool. and can initialize it to false on the same line as the declare

File: src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c [hunk 11]
```diff
@@ -30,41 +37,266 @@ const ActorInit En_Shopnuts_InitVars = {
...
+    } else {
+        Actor_SetHeight(&this->actor, 20.0f);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Shopnuts/func_80AFAF64.s")
+s32 EnShopnuts_OverrideLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot,
+                                void* thisx) {
+    EnShopnuts* this = THIS;
+    if ((limbIndex == 9) && (this->actionFunc == EnShopnuts_ThrowNut)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:257]: can put a line break after THIS

**Review by fig02 — COMMENTED** [review]
nice, a couple very minor things

**Review by fig02 — APPROVED** [review]

**Review by Roman971 — APPROVED** [review]

**Pull request merged**